src/middlewared/middlewared/plugins/enterprise/m_series_nvdimm_freebsd.py

```diff
@@ -0,0 +1,80 @@
+# -*- coding=utf-8 -*-
+import itertools
+import json
+import logging
+import os
+import struct
+
+import sysctl
+
+from middlewared.service import CallError, private, Service
+
+logger = logging.getLogger(__name__)
+
+
+class EnterpriseService(Service):
+
+    DATA = None
+    ERROR = "Data not retrieved yet"
+
+    @private
+    def setup_m_series_nvdimm(self):
+        try:
+            # This should be ran before pool import so we cache the value obtained on first middlewared startup
+            cache_path = "/tmp/m-series-nvdimm"
+            if os.path.exists(cache_path):
+                with open(cache_path) as f:
+                    self.DATA = json.load(f)
+                    return
+
+            result = []
+
+            for i in itertools.count(0):
+                try:
+                    pnpinfo = sysctl.filter(f"dev.nvdimm.{i}.%pnpinfo")[0].value
```

Comment:
Over talking with mav, its probably better if we use his flash tool to retrieve firmware version instead of sysctl.

Also he says we dont necessarily need to do it before pool import and it should work.